extensions/markdown/src/extension.ts

```diff
@@ -265,4 +266,33 @@ class MDDocumentContentProvider implements TextDocumentContentProvider {
 			}, 300);
 		}
 	}
+}
+
+class DocumentHeadingsProvider implements vscode.DocumentSymbolProvider {
+
+	// http://daringfireball.net/projects/markdown/syntax#header
+	private static _atxPattern = /^(#){1,6}\s+.+(\s+\1)?/;
```

Comment:
And headers can also indented up to three spaces for each indentation level. I'd make sure to handle the common case, a top level indent of between zero and three spaces (at four it becomes a code block). Handling headers in lists is probably not part of this feature.